### ISSUE TITLE
Feature request: the option to show the winning percentage of neural networks

### ISSUE THREAD

**Issue by ashinpan** [cutechess/cutechess#474]
At present, neural network engines like lc0 have to translate the winning probability of their moves into centipawns, and as all who have watched engine tournaments know very well, the translated results are not more than a joke. 

Therefore, I suggest that the gui version should give the engines like lc0 an option to show the winning percentages of their moves in original format. That is, the ability to say "55%" probability to win, rather than in centipawns.

**Comment by alwey** [comment]
We must be careful not to break protocols. We might be able to display a substitution of an engine's cp score if this has been defined (by the user!) for that engine.

**Comment by ashinpan** [comment]
Neural network engines usually give out the percentages of winning probabilities. It is for the sake of interfaces like cutechess that they have to translate the winning probability into centipawns.

**Comment by alwey** [comment]
The UCI interface communicates centipawns, not percentages.

**Comment by ashinpan** [comment]
Yes, but you can always implement this as a non-UCI optional feature so that neural engines can adopt it. Given the popularity of cutechess, this feature can even be incorporated into the next version of UCI specs.

**Comment by alwey** [comment]
It would take consensus in the community, esp. with people like Stefan Meyer-Kahlen or @HGMuller to extend the UCI interface. It would not help to mix things up and create incompatible dialects of UCI for standard chess. We have to be a bit careful.

**Comment by Prcuvu** [comment]
A neural network engine such as Leela Chess Zero does have its own conversion method from percentage to centipawns eval score. Maybe another neural network engine wants to implement a better conversion algorithm. Therefore it should not be a responsibility of cutechess.

I agree with @alwey adding a user-specified option to make cutechess display win percentage instead of centipawn score for the engine.

**Comment by ashinpan** [comment]
I cannot see how it can become incompatible given that this is only an addon to be optionally used by the user (i.e., engine), while everything else is true to UCI specs.

**Comment by ashinpan** [comment]
@Prcuvu  "I agree with @alwey adding a user-specified option to make cutechess display win percentage instead of centipawn score for the engine"---this is exactly my request. I am not asking cutechess to implement a conversion algorithm, but rather to remove the need for such a conversion.

**Comment by HGMuller** [comment]
I am skeptical that displaying percentages would solve any problem. The
conversion % -> cP done by LC0 will not be completely ridiculous, so if
the resulting cP are a joke, the percentage will be a joke too.

The undelying problem is that beyond a certain advantage a win gets 100%
certain. But to convert the win the score would need to be a measure of
progress, which it cannot be if it already scores 100% in the current game
position. So making a certain win the highest possible score is just as
bad as scoring a theoretical draw as 0 cP.

NN engines have to become more clever in this respect, and I am sure they
will. No need for GUI developers to panic in the mean time.

**Comment by ashinpan** [comment]
@HGMuller  Winning percentages and centipawns speak different languages. When an NN engine claims that it has "80%" probability of winning, we understand that it is possible, even probable if it is playing against a much stronger engine, that the remaining 20% will actually work towards the end, resulting in a draw or even a loss. In contrast, centipawns are much closer to the ground reality of material and positional advantages. If a top engine like Stockfish is 2 or more pawns ahead, we understand that its victory is almost guaranteed. This divergence between two concepts is why the conversion of winning percentage to centipawns appears like a joke. 

Therefore, IMHO, displaying percentages does solve a psychological problem for the viewers of an NN engine: it can tell them what to expect and what not to, in a much more accurate manner.

**Comment by HGMuller** [comment]
@ashinpan What you say only shows the conversion that is apparently used
is not realistic. If an advantage of 200cP is a certain win, a winning
chance of 80% should convert to ~120 cP. If centi-Pawn advantages would
not have a 1:1 correspondence with winning prospects, they would be
meaningless quantities to start with. But all conventional engines use
such a (sometimes game-phase-dependent) scoring to govern their play, and
they can be pretty strong when properly tuned.

**Comment by ashinpan** [comment]
@HGMuller  As far as I know, centipawns are calculated by traditional engines using the traditionally assigned piece values and the positions of the pieces on board. This concept itself does not exist with self-learning NN engines like lc0. 

On the other hand, the winning percentage of NN engines only mean, for instance: "If another copy of myself is playing against me, I have 80% probability of winning"---this is an evaluation of the whole board, rather than piecewise judgement. As viewers, we would like to see the actual judgement of NN engines, not a translated concept that they do not really have.

**Comment by HGMuller** [comment]
@ashinpan Piece values are just one term in an engine's evaluation, and
there is not much traditional about them. Usually they are tuned to give
the strongest play, and are not fixed, but dependent on the piece's
location on the board, and the game phase. Positional terms like Pawn
structure and King safety can contribute as much as 300 cP each. The
evaluation scores must represent winning prospects considering the
position as a whole, as striving for a high score is the only thing the
engine ever does.

So what you say is just not true; there is no fundamental difference
between a representation in terms of winning percentages or centi-Pawns.
They are the same thing, but expressed in a different, non-linearly
distorted scale.

**Comment by ashinpan** [comment]
@HGMuller  You said: "Piece values are just one term in an engine's evaluation, and there is not much traditional about them. Usually they are tuned to give the strongest play, and are not fixed, but dependent on the piece's location on the board, and the game phase. Positional terms like Pawn structure and King safety can contribute as much as 300 cP each. The evaluation scores must represent winning prospects considering the position as a whole, as striving for a high score is the only thing the engine ever does."

I have nothing to refute with above; I do agree that you are much better informed than me in this aspect. However: 

"So what you say is just not true; there is no fundamental difference between a representation in terms of winning percentages or centi-Pawns." 

I have to disagree with this. No one exactly knows how a NN engine arrives at a winning percentage; this is why neural networks are called black boxes. As far as self-learning NN engines are concerned, we can be sure that they are never explicitly taught such centipawn evaluations---this is human knowledge, not compatible with the zero approach, which NN engines have adopted.  

Anyhow, my request is from the viewers' perspective. We would like to see how exactly NN engines evaluate each stage of a game in their own native way.

**Comment by alwey** [comment]
It would be much cleaner to use the `info string` command (which cutechess currently ignores, BTW) to convey extra info. Transmitting cp 5000 for an even position may confuse the adjudicator (if a TCEC-style two-sided adjudication were used).

**Comment by ashinpan** [comment]
@alwey  I like your way better. It is best for an NN to avoid the centipawn concept if the latter is not actually used in engine calculations.

**Comment by rwbc** [comment]
Well, I thought you wanted to use it, instead of waiting for something which might never happen...
It seems your request was never urgent nor serious, otherwise you had figured out the uci option yourself.

Actually, I believe you did not even understand my answer, because the output is just a masking of percentages and has nothing to do with centipawns... I am out here.

**Comment by rwbc** [comment]
> 
> 
> It would be much cleaner to use the `info string` command (which cutechess currently ignores, BTW) to convey extra info. Transmitting cp 5000 for an even position may confuse the adjudicator (if a TCEC-style two-sided adjudication were used).

The info string would not be saved into the pgn though?

**Comment by alwey** [comment]
We do not save info strings currently. Currently cutechess saves games in a neat, concise format or in a minimal format.

**Comment by ashinpan** [comment]
@rwbc  I am not talking about playing games on my pc. Rather, for the sake of online tournaments like tcec. In such cases, we would like to see, e.g., "+56%" instead of the misleading "+56," which is available now.

**Comment by alwey** [comment]
@ashinpan : TCEC has its own version of cutechess in use, and the interface to the viewer is totally different, it is a web interface.
